File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    
    vector<vector<int>> graph(rows, vector<int>(columns));
    int num = 0;
    for(int i=0;i<rows;i++) {
        for(int j=0;j<columns;j++) {
            graph[i][j] = ++num;
        }
    }
    
    for(auto q: queries) {
        int x1 = q[0]-1, y1 = q[1]-1, x2 = q[2]-1, y2 = q[3]-1;
        int nx = x1, ny = y1;
        int prev = graph[nx][ny];
        int minval = rows * columns;
        
        // right
        while(ny+1 <= y2) {
            minval = min(prev, minval);
            int next = graph[nx][ny+1];
            graph[nx][ny+1] = prev;
            prev = next;
            ny++;
        }
        
        // down
        while(nx+1 <= x2) {
            minval = min(prev, minval);
            int next = graph[nx+1][ny];
            graph[nx+1][ny] = prev;
            prev = next;
            nx++;
        }
        
        // left
        while(ny-1 >= y1) {
            minval = min(prev, minval);
            int next = graph[nx][ny-1];
            graph[nx][ny-1] = prev;
            prev = next;
            ny--;
        }
        
        // up
        while(nx-1 >= x1) {
            minval = min(prev, minval);
            int next = graph[nx-1][ny];
            graph[nx-1][ny] = prev;
            prev = next;
            nx--;
        }
        answer.push_back(minval);
    }
    
    
    return answer;
}

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.swift
import Foundation

func solution(_ rows:Int, _ columns:Int, _ queries:[[Int]]) -> [Int] {
    
    var answer = [Int]()
    
    var grid = Array(repeating: Array(repeating: 0, count: columns), count: rows)
    var num = 1
    for i in 0..<rows {
        for j in 0..<columns {
            grid[i][j] = num
            num += 1
        }
    }
    
    let dirx = [0,1,0,-1]
    let diry = [1,0,-1,0]
    
    
    for query in queries {        
        let sx = query[0]-1
        let sy = query[1]-1
        let ex = query[2]-1
        let ey = query[3]-1
        
        var x = sx
        var y = sy
        
        var prev = grid[sx][sy]
        var d = 0
        var minVal = prev
        
        while true {
            minVal = min(minVal, prev)
            var nx = x + dirx[d]
            var ny = y + diry[d]
            if nx<sx || nx>ex || ny<sy || ny>ey {
                d += 1
                if d == 4 { break }
                nx = x + dirx[d]
                ny = y + diry[d]
            }
            let now = grid[nx][ny]
            grid[nx][ny] = prev
            prev = now
            
            x = nx
            y = ny
        }
        answer.append(minVal)
    }
    
    return answer
}